### ISSUE TITLE
Unable to source multiple integrations of the same type

### ISSUE THREAD

**Issue by philwhiteuk** [ewilde/terraform-provider-runscope#9]
The current implementation for the `runscope_integration` data source assumes that only one integration of each _type_ exists for a team. This causes an issue whereby, if the team has multiple integrations of the same type, the team can only bind test environments to the first one that gets returned in the api response.

**Comment by philwhiteuk** [comment]
Ideally I would like to be able to fetch back a list of integrations of a particular type and then filter the data source using Terraform's interpolation functions. However this would be a breaking change for anyone that has already written stuff around this data source.

I suggest instead, optionally, also supplying a _description_ and matching the integration against this?

https://github.com/ewilde/terraform-provider-runscope/pull/10

**Comment by ewilde** [comment]
@philwhiteuk thanks for spotting this and the PR . Could you post an example in HCL of the two proposals. We can make a breaking change if it’s going to be cleaner

**Comment by philwhiteuk** [comment]
@ewilde so this is what I envisaged

*Proposal 1: optionally matching against description*
```
data "runscope_integration" "pagerduty" {
  team_uuid = "xxx-xxx-xxx"
  type = "pagerduty"
}

data "runscope_integration" "slack_main" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
  description = "Slack: #main-channel"
}

data "runscope_integration" "slack_dev" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
  description = "Slack: #dev-channel"
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integration.pagerduty}",
    "${data.runscope_integration.slack_main}",
    "${data.runscope_integration.slack_dev}",
  ]
}
```

*Proposal 2: return list of integrations which can be filtered using interpolation*
```
data "runscope_integration" "pagerduty" {
  team_uuid = "xxx-xxx-xxx"
  type = "pagerduty"
}

data "runscope_integration" "slack" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integration.pagerduty[0]}",
    "${element(
        data.runscope_integration.slack, 
        index(data.runscope_integration.slack.*.Description, "Slack: #main-channel")
      )}",
    "${element(
        data.runscope_integration.slack, 
        index(data.runscope_integration.slack.*.Description, "Slack: #dev-channel")
      )}",
  ]
}
```

**Comment by ewilde** [comment]
@philwhiteuk I like the idea of filtering against description and/ or returning a list, think they both make sense. Looking at how the AWS provider has tackled this and iterating on your above proposal:

### Proposal 1: optionally matching against description using a [filter type set](https://github.com/terraform-providers/terraform-provider-aws/blob/b5bb97c94818f74ea720346063ad720c006aede1/aws/data_source_aws_ami.go#L19)
```hcl
data "runscope_integration" "pagerduty" {
  team_uuid = "xxx-xxx-xxx"
  type = "pagerduty"
}

data "runscope_integration" "slack_main" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
  filter {
    name   = "description"
    values = ["Slack: #main-channel"]
  }
}

data "runscope_integration" "slack_dev" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
  filter {
    name   = "description"
    values = ["Slack: #dev-channel"]
  }
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integration.pagerduty}",
    "${data.runscope_integration.slack_main}",
    "${data.runscope_integration.slack_dev}",
  ]
}
```

### Proposal 2: return list of integrations which can be filtered using interpolation
Create a new data resource `runscope_integrations` (pluralized) which returns a list similar to [zone](https://www.terraform.io/docs/providers/aws/d/availability_zone.html) and [zones](https://www.terraform.io/docs/providers/aws/d/availability_zones.html)

```hcl
data "runscope_integrations" "pagerduty" {
  team_uuid = "xxx-xxx-xxx"
  type = "pagerduty"
}

data "runscope_integrations" "slack" {
  team_uuid = "xxx-xxx-xxx"
  type = "slack"
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integration.pagerduty[0]}",
    "${element(
        data.runscope_integration.slack, 
        index(data.runscope_integration.slack.*.Description, "Slack: #main-channel")
      )}",
    "${element(
        data.runscope_integration.slack, 
        index(data.runscope_integration.slack.*.Description, "Slack: #dev-channel")
      )}",
  ]
}
```
As a bonus I think either or both of these should be backwards compatible

**Comment by philwhiteuk** [comment]
@ewilde agreed, I like this solution. 

In that case I will amend https://github.com/ewilde/terraform-provider-runscope/pull/10 to use the filter approach instead and will take a look at adding pluralised separately.

**Comment by philwhiteuk** [comment]
@ewilde I have modified https://github.com/ewilde/terraform-provider-runscope/pull/10 to incorporate filtering.

Tests passing 💯

**Comment by ewilde** [comment]
Closed by #10

**Comment by philwhiteuk** [comment]
@ewilde I have created a PR https://github.com/ewilde/terraform-provider-runscope/pull/16 for an `integrations` data source which, using the same filtering logic, returns a list of matching integration ids.

I had to change the schema of the `environment.integrations` key as previously it was a list of maps and there is no way in Terraform to dynamically add a list of maps from a data source. Specifically, our use-case was that we needed to conditionally add integrations and limitations around Terraform's interpolation syntax prevented this.

With the inclusion of this PR we could do the following:

```
data "runscope_integration" "pagerduty" {
  team_uuid = "xxx-xxx-xxx"
  type = "pagerduty"
}

data "runscope_integration" "slack" {
  team_uuid = "xxx-xxx-xxx"
  filter {
    name   = "type"
    values = ["slack"]
  }
  filter {
    name   = "description"
    values = ["Slack: #main-channel","Slack: #dev-channel"]
  }
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integration.pagerduty.id}",
    "${data.runscope_integrations.slack.ids}",
  ]
}
```

It also allows for the following:

```
data "runscope_integration" "doesnt_exist" {
  team_uuid = "xxx-xxx-xxx"
  filter {
    name   = "type"
    values = ["foobar"]
  }
}

resource "runscope_environment" "environment" {
  bucket_id = "${var.runscope["bucket_id"]}"
  name = "${var.runscope["environment"]}"
  integrations = [
    "${data.runscope_integrations.doesnt_exist.ids}",
  ]
}
```

Tests pass, let me know what you think!
